README.md
<!-- markdownlint-configure-file {
  "MD033": false,
  "MD041": false
} -->

# AWS Route 53 Dynamic DNS

[![Crates.io](https://img.shields.io/crates/l/r53-ddns)](https://github.com/a1ecbr0wn/r53-ddns/blob/main/LICENSE) [![Crates.io](https://img.shields.io/crates/v/r53-ddns)](https://crates.io/crates/r53-ddns) [![Build Status](https://github.com/a1ecbr0wn/r53-ddns/workflows/CI%20Build/badge.svg)](https://github.com/a1ecbr0wn/r53-ddns/actions/workflows/build.yml) [![dependency status](https://deps.rs/repo/github/a1ecbr0wn/r53-ddns/status.svg)](https://deps.rs/repo/github/a1ecbr0wn/r53-ddns) [![snapcraft.io](https://snapcraft.io/r53-ddns/badge.svg)](https://snapcraft.io/r53-ddns)

## Introduction

This app provides a way to keep a consistant url for a network where the external ip address may change from time to time, by adding a record to a domain where the primary DNS is hosted by the [Amazon Route 53](https://aws.amazon.com/route53/) service.

There are many reasons why you would want this, the most common is to provide a URL to a service hosted on a domestic network where the ISP is not providing a static ip address, e.g. webserver, vpn to home, etc..  There are other DDNS services out there that may give you a free option, I just want to use my own domain.

This is not is a publicly facing DDNS API, for that I would recomend another repository with a similar name: [aws-ddns](https://github.com/dixonwille/aws-ddns).  This is an application that provides DDNS using the R53 API, it is intended that you run this on one a computer within your network at a frequency that you are happy with.

## Pre-Requisites

- An AWS account, this tool uses the Amazon Route 53 service which is part of AWS, so you need an account
- Your own domain, the Route 53 service allows you to manage the DNS of your own domain and this tool allows you to map a subdomain to your current external IP address
- [AWS credentials](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-configure.html) permissioned to modify R53 records stored in a local credentials file located at `$HOME/.aws/credentials` as you would for the AWS CLI.  You can also use the `AWS_SHARED_CREDENTIALS_FILE` environment variable to locate your credential file, or use `AWS_ACCESS_KEY` / `AWS_SECRET_ACCESS_KEY` environment variables to specify your credentials.

## How it works

The application works gets it's external network address over https from some of the many services that are available online.  The application has a number of these sites set up as default, but you can choose your own by providing a comma separated list of sites using the `-i` parameter.  Many of these service have limits to the frequency that you can call them, so r532-ddns limits each run to check a random two services out of the services available.  The value returned for the ip address is compared to the value that is stored in the Amazon Route 53 DNS settings, and if they differ, the DNS recordmis updated in Route 53.

## Options

``` text
$ r53-ddns -h

Set an Amazon Route 53 DNS record for the server/network

Usage: r53-ddns [OPTIONS]

Options:
  -s, --subdomain <SUBDOMAIN>          The subdomain to save (required)
  -d, --domain <DOMAIN>                The domain to save the record in (required)
  -r, --region <REGION>                The aws region [default: us-east-1]
  -i, --ipaddress-svc <IPADDRESS_SVC>  The ip address services to use, e.g. ident.me,ifconfig.me/ip
  -n, --nat                            The record is a nat router and so a *.<subdomain>.<domain> CNAME record will be set
  -l, --logdir <LOGDIR>                Absolute path for the directory where log file should be written [default: /var/tmp]
  -a, --alert-script <ALERT_SCRIPT>    Script called on error or ip address changes, see: https://r53-ddns.a1ecbr0wn.com/alert-script
  -c, --check <CHECK>                  Consecutive check gap in seconds for continuous checking [default: 0]
  -v, --verbose                        Verbose logging
  -V, --version                        Print version information
  -h, --help                           Print help
```

- To set a DNS record, the application must be provided with the subdomain and domain parameters.
- If the region for the domain is in a different region to the default `us-east-1`, it needs to be specified with the `-r` parameter.
- The default list of ip address https services can be overridden with the `-i` parameter.
- If the `-n` parameter is supplied, an additional `*.<subdomain>.<domain>` DNS record is set.  This can be used to route traffic to all hosts with that pattern to the network.
- To get the application to work continuously, the `-c` parameter can be used to pass in the length of the gap in seconds between consecutive checks.  Some of the ip address web servers will return errors if they are called too often, this application tries to address this by randomising the services that are used, but it is recomended that with the default list, the consecutive check gap is not below 300 seconds.

## Installation

- [Homebrew](docs/install-homebrew.md)
- [Snap](docs/install-snapcraft.md)
- [Cargo](docs/install-cargo.md)
- [Other Package Managers](docs/install-other.md)

## Usage

r53-ddns can be used adhoc if you wish but you probably want to set this up to run continuously in case your external ip address changes.  There are two ways that you can do this, using a job scheduler such as cron or as a service.

- [Setup cron](docs/setup-cron.md)
- [Setup Systemd Service on Linux](docs/setup-systemd.md)
- [Setup Launchd Service on MacOS](docs/setup-launchd.md)

## Issues

If you are having issues with the execution of this application, start by enabling verbose logging with the `-v` parameter.  If you are trying to resolve an issue while running this in cron or as a service, logging is enabled by default to the file `/var/tmp/r53-ddns.log` which is a file that should be writable whichever user you are sunning the application under.

If the issues are not obvious or you think there is a bug, please raise an issue via [GitHub](https://github.com/a1ecbr0wn/homebrew-r53-ddns/issues).

## Contribute

Feel free to contribute I would be happy to take a look at any PRs raised.

[![Get it from the Snap Store](https://snapcraft.io/static/images/badges/en/snap-store-black.svg)](https://snapcraft.io/r53-ddns)

## Other Info

- [Release Notes](RELEASE-NOTES.md)


RELEASE-NOTES.md
# Release Notes

## v1.0.1

- Fix to issue [#26](https://github.com/a1ecbr0wn/r53-ddns/issues/26) where, after an interface failover on a multi-wan network, the cached reference to the AWS client cannot resolve the AWS API and fails to update the new DNS record

## v1.0.2

- Added the following new parameter for supplying a script that is called when a DNS record is changed or an error is detected [#25](https://github.com/a1ecbr0wn/r53-ddns/issues/25).  For further information see: [alert-script](https://r53-ddns.a1ecbr0wn.com/alert-script):

```text
-a, --alert-script <ALERT_SCRIPT>
```

## v1.0.3

- Documentation URL change


src/cli.rs
use clap::Parser;
use rusoto_core::Region;

/// Struct containing the parsed command line arguments
#[derive(Parser)]
#[command(name = "r53-ddns")]
#[command(bin_name = "r53-ddns")]
#[command(author, version, about, long_about = None, arg_required_else_help(true), disable_version_flag(true))]
pub struct Options {
    /// The subdomain to save (required)
    #[arg(short, long, value_parser, display_order(1))]
    pub subdomain: Option<String>,

    /// The domain to save the record in (required)
    #[arg(short, long, value_parser, display_order(2))]
    pub domain: Option<String>,

    /// The aws region
    #[arg(short, long, value_parser, display_order(3), default_value = Region::UsEast1.name())]
    pub region: String,

    /// The ip address services to use, e.g. ident.me,ifconfig.me/ip
    #[arg(short, long, value_parser, display_order(4))]
    pub ipaddress_svc: Option<String>,

    /// The record is a nat router and so a *.<subdomain>.<domain> CNAME record will be set
    #[arg(short, long, value_parser, display_order(5))]
    pub nat: bool,

    /// Absolute path for the directory where log file should be written
    #[arg(
        short,
        long,
        value_parser,
        display_order(6),
        default_value = "/var/tmp"
    )]
    pub logdir: String,

    /// Script called on error or ip address changes, see: https://r53-ddns.a1ecbr0wn.com/alert-script
    #[arg(short, long, value_parser, display_order(7))]
    pub alert_script: Option<String>,

    /// Consecutive check gap in seconds for continuous checking
    #[arg(short, long, value_parser, display_order(8), default_value_t = 0)]
    pub check: u64,

    /// Verbose logging
    #[arg(short, long, value_parser, default_value_t = false)]
    pub verbose: bool,

    /// Debug logging
    #[arg(short = 'D', long, value_parser, hide(true), default_value_t = false)]
    pub debug: bool,

    /// Print version information
    #[arg(short = 'V', long, value_parser)]
    pub version: bool,
}


src/main.rs
mod cli;

use std::env;
use std::net::IpAddr;
use std::path::Path;
use std::process::{Command, ExitStatus};
use std::str::FromStr;

use clap::Parser;
use futures::stream::{FuturesUnordered, StreamExt};
use lazy_static::lazy_static;
use log::{debug, error, info, warn, LevelFilter};
use log4rs::append::console::ConsoleAppender;
use log4rs::append::rolling_file::policy::compound::roll::fixed_window::FixedWindowRoller;
use log4rs::append::rolling_file::policy::compound::trigger::size::SizeTrigger;
use log4rs::append::rolling_file::policy::compound::CompoundPolicy;
use log4rs::append::rolling_file::RollingFileAppender;
use log4rs::config::{Appender, Config, Root};
use log4rs::encode::pattern::PatternEncoder;
use log4rs::filter::threshold::ThresholdFilter;
use rand::seq::SliceRandom;
use regex::Regex;
use reqwest::Client;
use rusoto_core::{Region, RusotoError};
use rusoto_route53::{
    Change, ChangeBatch, ChangeResourceRecordSetsRequest, ListHostedZonesRequest,
    ListResourceRecordSetsRequest, ResourceRecord, ResourceRecordSet, Route53, Route53Client,
};
use snapcraft::check_snap_home;
use tokio::{
    join, select,
    time::{sleep, Duration},
};

#[tokio::main(flavor = "current_thread")]
async fn main() -> Result<(), RusotoError<RusotoError<()>>> {
    lazy_static! {
        static ref DESCRIPTION: String = format!(env!("CARGO_PKG_VERSION"));
    };

    let options = cli::Options::parse();
    let check_freq = options.check;

    // set up logging
    let log_dir = Path::new(&options.logdir);
    let (log_file, log_roll) = if log_dir.is_dir() {
        let mut log_file = log_dir.to_path_buf();
        log_file.push("r53-ddns.log");
        let mut log_roll = log_dir.to_path_buf();
        log_roll.push("r53-ddns.{}.log");
        (
            log_file.display().to_string(),
            log_roll.display().to_string(),
        )
    } else {
        println!(
            "`{}' is not a valid path for the log file",
            log_dir.display()
        );
        (
            "/var/tmp/r53-ddns.log".to_string(),
            "/var/tmp/r53-ddns.{}.log".to_string(),
        )
    };

    let stdout = ConsoleAppender::builder()
        .encoder(Box::new(PatternEncoder::new("{m}{n}")))
        .build();
    let rolling_policy = CompoundPolicy::new(
        Box::new(SizeTrigger::new(1024 * 512)), // 512kb
        Box::new(FixedWindowRoller::builder().build(&log_roll, 4).unwrap()),
    );
    let to_file = RollingFileAppender::builder()
        .encoder(Box::new(PatternEncoder::new("{d} - {m}{n}")))
        .build(&log_file, Box::new(rolling_policy))
        .unwrap();
    let console_level = if options.debug {
        LevelFilter::Debug
    } else if options.verbose {
        LevelFilter::Info
    } else {
        LevelFilter::Warn
    };
    let config = Config::builder()
        .appender(
            Appender::builder()
                .filter(Box::new(ThresholdFilter::new(console_level)))
                .build("stdout", Box::new(stdout)),
        )
        .appender(Appender::builder().build("to_file", Box::new(to_file)))
        .build(
            Root::builder()
                .appender("stdout")
                .appender("to_file")
                .build(LevelFilter::Info),
        )
        .unwrap();
    let _handle = log4rs::init_config(config).unwrap();

    let alert_script: String = if let Some(alert_script) = options.alert_script {
        let alert_script_path = Path::new(&alert_script);
        if alert_script_path.is_file() {
            alert_script
        } else {
            println!("invalid alert script supplied: {alert_script}");
            "".to_string()
        }
    } else {
        println!("no alert script");
        "".to_string()
    };

    if env::var("AWS_SHARED_CREDENTIALS_FILE").is_err() {
        // if we are in a snap, rusoto will fail to read the credentials file from the $HOME/.aws/credential,
        // so set up that path but pointing to the real home rather than the snap home
        let (in_snap, home) = check_snap_home();
        if in_snap {
            if let Some(mut credentials_file) = home {
                credentials_file.push(".aws");
                credentials_file.push("credentials");
                if credentials_file.exists() {
                    env::set_var(
                        "AWS_SHARED_CREDENTIALS_FILE",
                        credentials_file.as_path().to_str().unwrap(),
                    );
                    debug!(
                        "within snap, AWS_SHARED_CREDENTIALS_FILE set to {:?}",
                        credentials_file
                    );
                }
            }
        }
    }

    // Get the options for the ddns check, and run it
    if options.subdomain.is_some() && options.domain.is_some() {
        let mut subdomain_name = options.subdomain.unwrap();
        let mut zone_name = options.domain.unwrap();
        if is_valid_hostname(&subdomain_name) {
            if !subdomain_name.ends_with('.') {
                subdomain_name += ".";
            }
            if !zone_name.ends_with('.') {
                zone_name += ".";
            }
            info!("subdomain:     {}", subdomain_name.clone());
            info!("domain:        {}", zone_name.clone());
            let region = match Region::from_str(options.region.as_str()) {
                Ok(region) => region,
                Err(_) => Region::UsEast1,
            };
            info!("region:        {}", region.name());
            let client = Route53Client::new(region.clone());
            let zone_id = get_zone_id(&client, &zone_name).await;
            info!("zone id:       {zone_id}");

            let nat = options.nat;
            let ipaddresses: Option<Vec<String>> = options
                .ipaddress_svc
                .map(|addrs| addrs.split(',').map(|x| x.to_string()).collect());

            if check_freq == 0 {
                ddns_check(
                    &client,
                    &zone_id,
                    &zone_name,
                    &subdomain_name,
                    &ipaddresses,
                    nat,
                    &alert_script,
                )
                .await;
            } else {
                loop {
                    let client = Route53Client::new(region.clone());
                    ddns_check(
                        &client,
                        &zone_id,
                        &zone_name,
                        &subdomain_name,
                        &ipaddresses,
                        nat,
                        &alert_script,
                    )
                    .await;
                    sleep(Duration::from_millis(1000 * check_freq)).await;
                }
            }
        } else {
            let err_msg = format!("invalid subdomain value: {subdomain_name}");
            warn!("{err_msg}\n");
            if !alert_script.is_empty() {
                let msg = format!("{{ \"type\": \"error\", \"msg\": \"{err_msg}\" }}");
                let _ = call_alert_script(&alert_script, &msg);
            }
        }
    } else if options.subdomain.is_some() || options.domain.is_some() {
        println!("r53-ddns v{}\n", DESCRIPTION.as_str());
        error!("subdomain and domain parameters need to be supplied together");
        return Ok(());
    }

    if options.version {
        println!("r53-ddns v{}", DESCRIPTION.as_str());
        return Ok(());
    }

    Ok(())
}

async fn ddns_check(
    client: &Route53Client,
    zone_id: &str,
    zone_name: &str,
    subdomain_name: &str,
    ipaddresses: &Option<Vec<String>>,
    nat: bool,
    alert_script: &str,
) {
    let dns_name = format!("{subdomain_name}{zone_name}");
    let external_ip_future = get_external_ip_address(ipaddresses);
    let dns_ip_future = get_dns_record(
        client,
        zone_id,
        zone_name,
        subdomain_name,
        "A",
        alert_script,
    );
    let (external_ip_address, dns_ip_address) = join!(external_ip_future, dns_ip_future);
    if let Some(dns_ip_address) = dns_ip_address {
        if dns_ip_address != external_ip_address
            && !dns_ip_address.is_empty()
            && !external_ip_address.is_empty()
        {
            warn!(
                "{dns_name} ip address has changed from {dns_ip_address} to {external_ip_address}"
            );
            if !alert_script.is_empty() {
                let msg = format!("{{ \"type\": \"ip-change\", \"dns\": \"{dns_name}\", \"old\": \"{dns_ip_address}\", \"new\": \"{external_ip_address}\" }}");
                let _ = call_alert_script(alert_script, &msg);
            }
            set_dns_record(
                client,
                zone_id,
                zone_name,
                subdomain_name,
                "A",
                &external_ip_address,
            )
            .await;
        }
    } else if !external_ip_address.is_empty() {
        warn!("{dns_name} ip address is {external_ip_address}");
        if !alert_script.is_empty() {
            let msg = format!("{{ \"type\": \"ip-new\", \"dns\": \"{dns_name}\",  \"new\": \"{external_ip_address}\" }}");
            let _ = call_alert_script(alert_script, &msg);
        }
        set_dns_record(
            client,
            zone_id,
            zone_name,
            subdomain_name,
            "A",
            &external_ip_address,
        )
        .await;
    }
    if nat {
        let nat_subdomain_name = "\\052.".to_string() + subdomain_name;
        let dns_nat_cname = get_dns_record(
            client,
            zone_id,
            zone_name,
            &nat_subdomain_name,
            "CNAME",
            alert_script,
        )
        .await;
        if dns_nat_cname.is_none()
            || (dns_nat_cname.is_some() && dns_nat_cname.clone().unwrap() != dns_name)
        {
            info!("{nat_subdomain_name}{zone_name} nat CNAME set to {dns_name}");
            set_dns_record(
                client,
                zone_id,
                zone_name,
                &nat_subdomain_name,
                "CNAME",
                dns_name.as_str(),
            )
            .await
        }
    }
}

/// This function checks to see whether the subdomain_name entered into the zone and cname tags is a valid subdomain_name.
fn is_valid_hostname(subdomain_name: &str) -> bool {
    if subdomain_name.len() > 255 {
        return false;
    }
    let re = Regex::new(r"^[a-zA-Z\d-]{1,63}$").unwrap();
    for s in subdomain_name.split('.') {
        if !re.is_match(s) && !s.is_empty() {
            return false;
        }
    }
    true
}

/// Returns the zone id for a given zone name
async fn get_zone_id(client: &Route53Client, zone_name: &str) -> String {
    let mut next_marker: Option<String> = None;
    loop {
        // Create a request to list all hosted zones
        let request = ListHostedZonesRequest {
            marker: next_marker,
            ..Default::default()
        };
        // Send the request and print the response
        if let Ok(response) = client.list_hosted_zones(request).await {
            for zone in response.hosted_zones {
                if zone.name == zone_name {
                    let zone_id = zone.id.rsplit_once('/').unwrap().1.to_string();
                    return zone_id;
                }
            }
            if response.is_truncated {
                next_marker = response.next_marker;
            } else {
                return "".to_string();
            }
        } else {
            return "".to_string();
        }
    }
}

/// Get the external ip address from an external service
async fn get_external_ip_address(ipaddresses: &Option<Vec<String>>) -> String {
    let mut futures = FuturesUnordered::new();

    let default_ipaddresses: Vec<String> = [
        "ident.me",
        "ifconfig.me/ip",
        "icanhazip.com",
        "myexternalip.com/raw",
        "ipecho.net/plain",
        "checkip.amazonaws.com",
        "myip.dnsomatic.com",
    ]
    .iter()
    .map(|x| x.to_string())
    .collect();
    let ipaddresses: Vec<String> = match ipaddresses {
        Some(ipaddresses) => {
            if ipaddresses.len() > 1 {
                ipaddresses
                    .choose_multiple(&mut rand::thread_rng(), 2)
                    .map(|x| x.to_string())
                    .collect()
            } else {
                default_ipaddresses
                    .choose_multiple(&mut rand::thread_rng(), 2)
                    .map(|x| x.to_string())
                    .collect()
            }
        }
        None => default_ipaddresses
            .choose_multiple(&mut rand::thread_rng(), 2)
            .map(|x| x.to_string())
            .collect(),
    };

    // let addresses_fut = ipaddresses.into_iter().map(|x| get_http_resp(x.as_str()));
    for ipaddress in ipaddresses {
        let address_fut = get_http_resp(ipaddress);
        futures.push(address_fut);
    }
    for _ in 0..futures.len() {
        select! {
            response = futures.select_next_some() => {
                if let Ok((external_ip_address, external_address_svc)) = response {
                    info!("external ip:   {}  (from {})", external_ip_address, external_address_svc);
                    return external_ip_address;
                }
            }
        }
    }
    "".to_string()
}

async fn get_http_resp(address: String) -> Result<(String, String), ()> {
    let client = Client::new();
    if let Ok(resp) = client.get(format!("https://{address}")).send().await {
        if let Ok(response) = resp.text().await {
            let response = response.trim().to_string();
            if response.parse::<IpAddr>().is_ok() {
                return Ok((response, address.to_string()));
            }
        }
    }
    Err(())
}

//////////////////////////////
// Amazon Route 53 Interaction
//////////////////////////////

/// Lists the ip address of a given zone/host A record
async fn get_dns_record(
    client: &Route53Client,
    zone_id: &str,
    zone_name: &str,
    subdomain_name: &str,
    record_type: &str,
    alert_script: &str,
) -> Option<String> {
    let dns_name = format!("{subdomain_name}{zone_name}");
    let request = ListResourceRecordSetsRequest {
        hosted_zone_id: zone_id.to_string(),
        start_record_name: Some(dns_name.clone()),
        ..Default::default()
    };
    match client.list_resource_record_sets(request).await {
        Ok(response) => {
            let record_sets = response.resource_record_sets;
            for record_set in record_sets {
                if record_set.name == dns_name && record_set.type_ == *record_type {
                    if let Some(records) = &record_set.resource_records {
                        if let Some(record) = records.first() {
                            let ip_record = record.value.clone();
                            if record_type == "A" {
                                if ip_record.parse::<IpAddr>().is_ok() {
                                    info!("dns ip:        {ip_record}");
                                    return Some(ip_record);
                                }
                            } else {
                                info!("dns ip:        {ip_record}");
                                return Some(ip_record);
                            }
                        }
                    }
                }
            }
            debug!("No record for {dns_name} currently set up in Route 53")
        }
        Err(x) => {
            let err_msg = format!(
                "Unable to retrieve the current dns address for {dns_name}: {x} Home={}",
                env::var("AWS_SHARED_CREDENTIALS_FILE").unwrap()
            );
            warn!("{err_msg}");
            if !alert_script.is_empty() {
                let msg = format!("{{ : \"error\", \"msg\": \"{err_msg}\" }}");
                let _ = call_alert_script(alert_script, &msg);
            }
        }
    }
    None
}

/// Creates resource records in the given hosted zone
async fn set_dns_record(
    client: &Route53Client,
    zone_id: &str,
    zone_name: &str,
    subdomain_name: &str,
    record_type: &str,
    record_value: &str,
) {
    // Build the request to change the resource record sets
    let dns_name = format!("{subdomain_name}{zone_name}");
    let request = ChangeResourceRecordSetsRequest {
        hosted_zone_id: zone_id.to_string(),
        change_batch: ChangeBatch {
            changes: vec![Change {
                action: String::from("UPSERT"),
                resource_record_set: ResourceRecordSet {
                    name: dns_name,
                    resource_records: Some(vec![ResourceRecord {
                        value: record_value.to_string(),
                    }]),
                    ttl: Some(300),
                    type_: record_type.to_string(),
                    ..Default::default()
                },
            }],
            comment: Some(String::from("Updated by r53-ddns")),
        },
    };

    // Call the Route53 API to change the resource record sets
    let _response = client.change_resource_record_sets(request).await;
}

pub fn call_alert_script(script: &str, message: &str) -> std::io::Result<ExitStatus> {
    Command::new(script).args([message]).spawn()?.wait()
}
